### PR TITLE
AGW: mobilityD: Add API for gwinfo

### PR DESCRIPTION
**Pull request by pshelar**
## Summary

This adds API for GW info in mobilityD. This would be used by pipelineD.
Rather than accessing redis store directly from pipelineD, this patch adds API that can be called from pipelineD. 

## Test Plan

ran `make test` and `make integ_test` on local setup. 

## Additional Information
Not a change is backwards-breaking

### CODE REVIEW

**Comment by ulaskozat** [comment]
@pshelar could you add more context into the summary?

**Comment by ulaskozat** [comment]
typo in summary: pipilineD

File: lte/gateway/python/magma/mobilityd/uplink_gw.py [hunk 1]
```diff
@@ -60,6 +60,10 @@ def mac_exists(self) -> str:
         return DHCP_Router_Mac_key in self._backing_map
 
     def update_mac(self, value: str):
+        # TODO: enhance check for MAC address sanity.
+        if value.find(':') == -1:
+            return
+
```
Comment (ulaskozat) [inline lte/gateway/python/magma/mobilityd/uplink_gw.py]: check with Nick as he may already have a function that checks MAC address sanity.

File: lte/protos/mobilityd.proto [hunk 2]
```diff
@@ -154,4 +160,10 @@ service MobilityService {
   // allocated from them. If force is set, then will remove all IP blocks,
   // regardless of whether any IPs have been allocated.
   rpc RemoveIPBlock (RemoveIPBlockRequest) returns (RemoveIPBlockResponse);
+
+  // Get default GW mac adresss
```
Comment (ulaskozat) [inline lte/protos/mobilityd.proto]: GW -> Internet Gateway
adresss -> adress

This returns both mac and IPv4 addresses.

File: lte/protos/mobilityd.proto [hunk 3]
```diff
@@ -154,4 +160,10 @@ service MobilityService {
...
+  rpc GetGatewayInfo (magma.orc8r.Void) returns (GWInfo);
+
+  // Set ip and mac address of def gw
```
Comment (ulaskozat) [inline lte/protos/mobilityd.proto]: def gw -> default Internet Gateway

File: lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py [hunk 4]
```diff
@@ -329,6 +330,30 @@ def test_get_ip_for_unknown_subscriber(self):
         self.assertEqual(err.exception.code(),
                          grpc.StatusCode.NOT_FOUND)
 
+    def test_get_set_gw_info(self):
```
Comment (ulaskozat) [inline lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py]: why "set" is in the function name? Looks like we are getting the gateway info from mobilityd and from shell, and compare....

File: lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py [hunk 5]
```diff
@@ -329,6 +330,30 @@ def test_get_ip_for_unknown_subscriber(self):
...
+        gw_info = self._stub.GetGatewayInfo(Void())
+        gw_ip_get = str(ipaddress.ip_address(gw_info.ip.address))
+
+        def_gw_cmd = "ip route show |grep default| awk '{print $3}'"
+        p = subprocess.Popen([def_gw_cmd], stdout=subprocess.PIPE, shell=True)
+        def_ip = p.stdout.read().decode("utf-8").strip()
+        self.assertEqual(def_ip, gw_ip_get)
+
+    def test_set_set_gw_info(self):
```
Comment (ulaskozat) [inline lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py]: why "_set_set"?

Comment (pshelar) [reply lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py]: fixed.

File: lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py (hunk 4 above)
Comment (pshelar) [reply lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py]: fixed

File: lte/protos/mobilityd.proto (hunk 3 above)
Comment (pshelar) [reply lte/protos/mobilityd.proto]: fixed.

File: lte/gateway/python/magma/mobilityd/uplink_gw.py (hunk 1 above)
Comment (pshelar) [reply lte/gateway/python/magma/mobilityd/uplink_gw.py]: I have plan to use netifaces mac object as TODO.

File: lte/protos/mobilityd.proto [hunk 6]
```diff
@@ -110,6 +110,13 @@ message RemoveIPBlockResponse {
   repeated IPBlock ip_blocks = 1;
 }
 
+// This msg is used for get/set API for mobilitiD UplinkGatewayInfo object
```
Comment (ardzoht) [inline lte/protos/mobilityd.proto]: typo: mobilityd

File: lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py [hunk 7]
```diff
@@ -329,6 +330,30 @@ def test_get_ip_for_unknown_subscriber(self):
         self.assertEqual(err.exception.code(),
                          grpc.StatusCode.NOT_FOUND)
 
+    def test_get_gw_info(self):
+        gw_info = self._stub.GetGatewayInfo(Void())
+        gw_ip_get = str(ipaddress.ip_address(gw_info.ip.address))
+
+        def_gw_cmd = "ip route show |grep default| awk '{print $3}'"
+        p = subprocess.Popen([def_gw_cmd], stdout=subprocess.PIPE, shell=True)
```
Comment (ardzoht) [inline lte/gateway/python/magma/mobilityd/tests/rpc_servicer_tests.py]: I'd recommend always setting shell to False, even though it's a unit test, if we can avoid (shell=True) the better. You can use `shlex.split()` to get a list of params and use that to accomplish the same.

File: lte/gateway/python/magma/mobilityd/uplink_gw.py [hunk 8]
```diff
@@ -60,6 +60,10 @@ def mac_exists(self) -> str:
         return DHCP_Router_Mac_key in self._backing_map
 
     def update_mac(self, value: str):
+        # TODO: enhance check for MAC address sanity.
+        if value.find(':') == -1:
```
Comment (ardzoht) [inline lte/gateway/python/magma/mobilityd/uplink_gw.py]: `if not ':' in value:`

**Review by ardzoht — CHANGES_REQUESTED** [review]

**Comment by ardzoht** [comment]
insync-checkin is failing due to mobilityd cloud protos not being updated, can you please regenerate them to fix this?

File: lte/protos/mobilityd.proto [hunk 9]
```diff
@@ -154,4 +161,10 @@ service MobilityService {
   // allocated from them. If force is set, then will remove all IP blocks,
   // regardless of whether any IPs have been allocated.
   rpc RemoveIPBlock (RemoveIPBlockRequest) returns (RemoveIPBlockResponse);
+
+  // Get default Internet Gateway mac addresss
```
Comment (ulaskozat) [inline lte/protos/mobilityd.proto:165]: IP and mac address

**Pull request closed without merging**